Sources/SwiftWarningControl/WarningControlRegions.swift

```diff
@@ -0,0 +1,313 @@
+  ) {
+    self.bounds = (start, end)
+    self.regions = [:]
+    self.parent = parent
+  }
+
+  /// Check if this node's bounds contain a location
+  /// The regions are `[start, end)`, meaning the region's
+  /// starting point is included in the interval, but the end point
+  /// is not. This is done in order to support fully-adjacent regions,
+  /// e.g. two function declarations which are alongside each other with
+  /// a `@warn` attribute on each will result in two warning control regions
+  /// where the first region ends at the exact same `AbsolutePosition` as
+  /// the start of the next region.
+  func contains(_ location: AbsolutePosition) -> Bool {
+    bounds.start <= location && location < bounds.end
+  }
+
+  /// Check if this node's bounds contain another node
+  func contains(_ node: WarningControlRegionNode) -> Bool {
+    bounds.start <= node.bounds.start && node.bounds.end <= bounds.end
+  }
+
+  /// Check if this node's bounds match another node's
+  func hasSameBoundaries(with node: WarningControlRegionNode) -> Bool {
+    bounds.start == node.bounds.start && bounds.end == node.bounds.end
+  }
+
+  /// Add a region with the same bounds as this node
+  func addRegion(_ region: WarningControlRegion) {
+    precondition(bounds.start == region.bounds.start && bounds.end == region.bounds.end)
+    regions[region.diagnosticGroupIdentifier] = region
+  }
+
+  /// Get region with specific identifier if it exists
+  func getRegion(for diagnosticGroupIdentifier: String) -> WarningControlRegion? {
+    return regions[diagnosticGroupIdentifier]
+  }
+
+  /// Add a child region that is directly nested within this region
+  func addChild(_ node: WarningControlRegionNode) {
+    precondition(contains(node))
+    children.append(node)
+    node.parent = self
+    children.sort()
+  }
+}
+
+extension WarningControlRegionNode {
+  /// Find the most deeply-nested warning control region with a given diagnostic group identifier
+  /// containing the specified location.
+  ///
+  /// Lookup complexity is `O(log n + d)` where `n` is the branching factor and `d` is the warning
+  /// control node depth of the queried position.
+  func innermostContainingRegion(
+    at location: AbsolutePosition,
+    for diagnosticGroupIdentifier: String
+  ) -> WarningControlRegion? {
+    precondition(contains(location))
+    if let childRegionContainingLocation = findChildContaining(location),
+      let nestedResult = childRegionContainingLocation.innermostContainingRegion(
+        at: location,
+        for: diagnosticGroupIdentifier
+      )
+    {
+      return nestedResult
+    }
+    // If no child region has this warning group, check self
+    return getRegion(for: diagnosticGroupIdentifier)
+  }
+
+  /// Find the exact child node containing the specified interval node, or `nil` if one is not found.
+  func findChildContaining(_ node: WarningControlRegionNode) -> WarningControlRegionNode? {
+    precondition(contains(node))
```

Comment:
I don’t think this function matches the description. The comment says that it matches by interval but it only looks up by start location. FWIW, I think we don’t really need this anyway and could just call into `findChildContaining(_:AbsolutePosition)`